scipy/stats/_qmc.py

```diff
@@ -481,6 +485,186 @@ def _perturb_discrepancy(sample: np.ndarray, i1: int, i2: int, k: int,
+    new_sample = np.empty_like(sample)
+
+    voronoi = spatial.Voronoi(sample)
+
+    for ii, idx in enumerate(voronoi.point_region):
+        # the region is a series of indices into self.voronoi.vertices
+        # remove point at infinity, designated by index -1
+        region = [i for i in voronoi.regions[idx] if i != -1]
+
+        # get the vertices for this region
+        verts = voronoi.vertices[region]
+
+        # clipping would be wrong, we need to intersect
+        # verts = np.clip(verts, 0, 1)
+
+        # move sample towards centroids:
+        # Centroid in n-D is the mean for uniformly distributed nodes
+        # of a geometry.
+        centroid = np.mean(verts, axis=0)
+        new_sample[ii] = sample[ii] + (centroid - sample[ii]) * decay
+
+    # only update sample to centroid within the region
+    is_valid = np.all(np.logical_and(new_sample >= 0, new_sample <= 1), axis=1)
+    sample[is_valid] = new_sample[is_valid]
+
+    return sample
+
+
+def lloyd_centroidal_voronoi_tessellation(
+        sample: npt.ArrayLike,
+        *,
+        n_iters: IntNumber = 10,
+        seed: SeedType = None
+) -> np.ndarray:
+    """Centroidal Voronoi Tessellation.
+
+    Perturb a sample of points in :math:`[0, 1]^d` using Lloyd's algorithm.
+
+    Parameters
+    ----------
+    sample : array_like (n, d)
+        The sample to iterate on. With ``n`` the number of sample and ``d``
+        the dimension.
+    n_iters : int, optional
+        Number of iterations. Too many iterations tend to cluster the sample
+        as a hypersphere. Default is 10.
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance then that instance is
+        used.
+
+    Returns
+    -------
+    sample : array_like (n, d)
+        The sample after `n_iters` of Lloyd's algorithm.
+
+    Notes
+    -----
+    Lloyd's algorithm is an iterative process which purpose is to improve
+    the dispersion of a sample. For a given sample: (i) compute a Voronoi
+    Tessellation; (ii) find the centroid of each Voronoi cell; (iii) move the
+    samples toward the centroid of their respective cell.
+
+    The process converges to equally spaced points. It implies that measures
+    like the discrepancy could suffer from too many iterations. On the other
+    hand, L1 and L2 distances should improve. This is especially true with
+    QMC methods which tend to favor the discrepancy over other criteria.
+
+    .. warning::
+
+       The Voronoi Tessellation step is expensive and quickly becomes
+       intractable with dimensions as low as 10 even for a sample of size
+       as low as 1000.
+
+    References
+    ----------
+    .. [1] Lloyd. "Least Squares Quantization in PCM".
+       IEEE Transactions on Information Theory, 1982.
```

Comment:
I noticed that sometimes it is referred to as the [Lloyd-Max algorithm](https://en.wikipedia.org/wiki/Lloyd%27s_algorithm#History). I wonder if we should give both of the authors who independently came up with the algorithm some credit.

Reply:
Sure, I will add this other reference.

Reply:
Both references are here, now, and it's described as the "Lloyd-Max algorithm". There are still a few places where "Lloyd's algorithm" is used, though. Consider whether these should be changed for consistency.